include/mapbox/std/weak.hpp

```diff
@@ -20,14 +20,26 @@ class WeakPtrSharedData {
     WeakPtrSharedData() = default;
 
     void sharedLock() {
-        assert(valid());
-        sharedLocks_++;
+        std::size_t noLocks = sharedLocks_;
+        std::size_t incremented;
+        do {
+            if (noLocks == kInvalidValue) return;
+            incremented = noLocks + 1;
+            // `compare_exchange_weak()` is invoked in a cycle to handle the case,
+            // if another thread has just modified `sharedLocks_` (so that it is not
+            // equal to `noLocks` any more). We early return, if `sharedLocks_` became
+            // equal to `kInvalidValue`.
+        } while (!sharedLocks_.compare_exchange_weak(noLocks, incremented));
     }
 
     void sharedUnlock() {
-        assert(valid());
-        assert(sharedLocks_ > 0u);
-        sharedLocks_--;
+        std::size_t noLocks = sharedLocks_;
+        std::size_t decremented;
+        do {
+            if (noLocks == kInvalidValue) return;
+            decremented = noLocks - 1;
+            assert(decremented >= 0u);
```

Comment:
This is always true, since `size_t` is unsigned.

Reply:
right, shall be `assert(noLocks > 0u);`

Reply:
Done